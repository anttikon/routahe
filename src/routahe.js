import moment from 'moment'
import { maxBy, minBy, get } from 'lodash'
import { gray, green, cyan, blueBright, yellow, magenta } from 'chalk'
import Table from 'cli-table'

const spacer = '  |'

const tableStyle = {
  chars: {
    top: '',
    'top-mid': '',
    'top-left': '',
    'top-right': '',
    bottom: '',
    'bottom-mid': '',
    'bottom-left': '',
    'bottom-right': '',
    left: '',
    'left-mid': '',
    mid: '',
    'mid-mid': '',
    right: '',
    'right-mid': '',
    middle: ' ',
  },
  style: { 'padding-left': 0, 'padding-right': 0 },
}

export const getRowProps = mode => {
  switch (mode) {
    case 'WALK':
      return { emoji: String.fromCodePoint(0x1f6b6), color: gray, name: mode }
    case 'RAIL':
      return { emoji: String.fromCodePoint(0x1f685), color: magenta, name: mode }
    case 'BUS':
      return { emoji: String.fromCodePoint(0x1f68c), color: blueBright, name: mode }
    case 'TRAM':
      return { emoji: String.fromCodePoint(0x1f683), color: green, name: mode }
    case 'SUBWAY':
      return { emoji: String.fromCodePoint(0x1f687), color: yellow, name: 'METRO' }
    case 'FERRY':
      return { emoji: String.fromCodePoint(0x1f6a2), color: cyan, name: mode }
    default:
      return { emoji: String.fromCodePoint(0x1f937), color: value => value, name: mode }
  }
}

export const getRouteHeaderRow = route => {
  const startTime = moment(minBy(route.legs, 'startTime').startTime)
  const endTime = moment(maxBy(route.legs, 'endTime').endTime)

  return `\n${startTime.format('HH:mm')} - ${endTime.format('HH:mm')} ${endTime.diff(startTime, 'minutes')}min`
}

export const getRowByLeg = leg => {
  const { color, emoji, name } = getRowProps(leg.mode)

  return [
    color(spacer),
    `${moment(leg.startTime).format('HH:mm')} - ${moment(leg.endTime).format('HH:mm')}`,
    `${emoji} ${color(name)}`,
    get(leg, 'route.shortName', ''),
    '->',
    leg.to.name,
  ]
}

export const getRouteRows = route => {
  const table = new Table(tableStyle)
  route.legs.forEach(leg => table.push(getRowByLeg(leg)))
  return table.toString()
}

export const getTripRow = (addressFrom, addressTo) => `${addressFrom.label} ${gray('->')} ${addressTo.label}`
