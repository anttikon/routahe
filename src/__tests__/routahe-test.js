import { getTripRow, getRouteHeaderRow, getRouteRows, getRowProps, getRowByLeg } from '../routahe'

const dummyRoute = {
  duration: 3279,
  legs: [
    {
      startTime: 1556046104000,
      endTime: 1556046240000,
      duration: 136,
      distance: 156.595,
      mode: 'WALK',
      from: { name: 'Origin' },
      to: { name: 'Kapteeninpuistikko' },
      route: null,
    },
    {
      startTime: 1556046240000,
      endTime: 1556046600000,
      duration: 360,
      distance: 1557.2188592626867,
      mode: 'BUS',
      from: { name: 'Kapteeninpuistikko' },
      to: { name: 'Lasipalatsi' },
      route: { shortName: '24' },
    },
    {
      startTime: 1556046600000,
      endTime: 1556046823000,
      duration: 223,
      distance: 289.68399999999997,
      mode: 'WALK',
      from: { name: 'Lasipalatsi' },
      to: { name: 'Rautatientori' },
      route: null,
    },
    {
      startTime: 1556046960000,
      endTime: 1556047980000,
      duration: 1020,
      distance: 11906.962915531054,
      mode: 'SUBWAY',
      from: { name: 'Rautatientori' },
      to: { name: 'Niittykumpu' },
      route: { shortName: 'M1' },
    },
    {
      startTime: 1556047980000,
      endTime: 1556048350000,
      duration: 370,
      distance: 258.101,
      mode: 'WALK',
      from: { name: 'Niittykumpu' },
      to: { name: 'Niittykumpu (M)' },
      route: null,
    },
    {
      startTime: 1556048940000,
      endTime: 1556049120000,
      duration: 180,
      distance: 1171.1100194427006,
      mode: 'BUS',
      from: { name: 'Niittykumpu (M)' },
      to: { name: 'Parolantie' },
      route: { shortName: '118' },
    },
    {
      startTime: 1556049120000,
      endTime: 1556049383000,
      duration: 263,
      distance: 312.56600000000003,
      mode: 'WALK',
      from: { name: 'Parolantie' },
      to: { name: 'Destination' },
      route: null,
    },
  ],
}

describe('getRowProps', () => {
  it('should return correct row props with WALK', () => {
    expect(getRowProps('WALK')).toMatchSnapshot()
  })
  it('should return default row props with ASDF', () => {
    expect(getRowProps('ASDF')).toMatchSnapshot()
  })
})
describe('getRouteHeaderRow', () => {
  it('should return route header row', () => {
    expect(getRouteHeaderRow(dummyRoute)).toMatchSnapshot()
  })
})
describe('getRowByLeg', () => {
  it('should return row containing route leg information', () => {
    expect(getRowByLeg(dummyRoute.legs[0])).toMatchSnapshot()
  })
})
describe('getRouteRows', () => {
  it('should return route segments as rows', () => {
    expect(getRouteRows(dummyRoute)).toMatchSnapshot()
  })
})
describe('getTripRow', () => {
  it('should return row containing trip information', () => {
    expect(getTripRow({ label: 'Tehtaankatu 24' }, { label: 'Korvatunturi' })).toMatchSnapshot()
  })
})
