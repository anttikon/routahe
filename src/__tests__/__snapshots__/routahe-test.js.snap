// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRouteHeaderRow should return route header row 1`] = `
"
22:01 - 22:56 54min"
`;

exports[`getRouteRows should return route segments as rows 1`] = `
"[90m  |[39m[90m [39m22:01 - 22:04[90m [39mðŸš¶ [90mWALK[39m [90m [39m   [90m [39m->[90m [39mKapteeninpuistikko
[94m  |[39m[90m [39m22:04 - 22:10[90m [39mðŸšŒ [94mBUS[39m  [90m [39m24 [90m [39m->[90m [39mLasipalatsi       
[90m  |[39m[90m [39m22:10 - 22:13[90m [39mðŸš¶ [90mWALK[39m [90m [39m   [90m [39m->[90m [39mRautatientori     
[33m  |[39m[90m [39m22:16 - 22:33[90m [39mðŸš‡ [33mMETRO[39m[90m [39mM1 [90m [39m->[90m [39mNiittykumpu       
[90m  |[39m[90m [39m22:33 - 22:39[90m [39mðŸš¶ [90mWALK[39m [90m [39m   [90m [39m->[90m [39mNiittykumpu (M)   
[94m  |[39m[90m [39m22:49 - 22:52[90m [39mðŸšŒ [94mBUS[39m  [90m [39m118[90m [39m->[90m [39mParolantie        
[90m  |[39m[90m [39m22:52 - 22:56[90m [39mðŸš¶ [90mWALK[39m [90m [39m   [90m [39m->[90m [39mDestination       "
`;

exports[`getRowByLeg should return row containing route leg information 1`] = `
Array [
  "[90m  |[39m",
  "22:01 - 22:04",
  "ðŸš¶ [90mWALK[39m",
  "",
  "->",
  "Kapteeninpuistikko",
]
`;

exports[`getRowProps should return correct row props with WALK 1`] = `
Object {
  "color": [Function],
  "emoji": "ðŸš¶",
  "name": "WALK",
}
`;

exports[`getRowProps should return default row props with ASDF 1`] = `
Object {
  "color": [Function],
  "emoji": "ðŸ¤·",
  "name": "ASDF",
}
`;

exports[`getTripRow should return row containing trip information 1`] = `"Tehtaankatu 24 [90m->[39m Korvatunturi"`;
